File: src/seq.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <math.h>

//matrix: input 
//sel consens
//n number of consens
void cons(int *matrix, int *len1, int *len2, int *sel, int *n, int *katMax) {
    printf("matrix: %dx%d\n", len1[0], len2[0]);

    int i, j, key;    
    for (i=0; i < len2[0]; i++) {    
	int count[katMax[0]+1];    
	for (j=0; j<katMax[0]+1; j++) {    
	    count[j] = 0;    
	}    

	//Test per column    
	for (j=0; j < len1[0]; j++) {     
	    key = matrix[i*len1[0]+j];    
	    count[key] = count[key]+1;    
	}    

	//Get most freq       
	sel[i] = count[0];    
	n[i] = 0;    
	for (j=1; j<katMax[0]+1; j++) {    
	    if (sel[i] < count[j]) {    
		sel[i] = count[j];    
		n[i] =j;    
	    }    
	}    
    }
}

void mtchSub(int *len, char **pattern, char **string, int *seqN, int *posM) {    
    char *pch = strstr(*string, *pattern);    
    //printf("Pattern: %s\n", pattern[0]);    
                              
    if(pch) {    
        // get first occurence         
        int pos = -1*(*string - pch);    
	posM[0] = pos;
        //printf("found @ %d\n", pos);    
        int i;                             
        //printf("len %d\n", len[0]);      
        for (i=pos; i<len[0]; i++) {                    
            char *c = string[0];        
            c = c+i;    
            if (c[0] == 'A') {    
                seqN[i] = 1;      
            } else if (c[0] =='C') {    
                seqN[i] = 2;    
            } else if (c[0] =='T') {    
                seqN[i] = 4;    
            } else if (c[0] =='G') {    
                seqN[i] = 3;    
            }    
        }        
    }        
}    

void mtchSubRev(int *len, char **pattern, char **string, int *seqN, int *posM) {    
    char *pch = strstr(*string, *pattern);    
    //printf("Pattern: %s\n", pattern[0]);    
                              
    if(pch) {    
        // get first occurence         
        int pos = -1*(*string - pch);    
	posM[0] = pos;
        //printf("found @ %d\n", pos);    
        int i;                             
        //printf("len %d\n", len[0]);      
        for (i=0; i<pos+len[0]; i++) {                    
            char *c = string[0];        
            c = c+i;    
            if (c[0] == 'A') {    
                seqN[i] = 1;      
            } else if (c[0] =='C') {    
                seqN[i] = 2;    
            } else if (c[0] =='T') {    
                seqN[i] = 4;    
            } else if (c[0] =='G') {    
                seqN[i] = 3;    
            }    
        }        
    }        
}    

void mtch(char **pattern, char **string, int *res) {
    char *pch = strstr(*string, *pattern);
    if(pch) { 
	res[0]=1 ;
    }
}

File: R/seq.R
#' @title Most frequent number per col of integer matrix
#' 
#' @description get large matrix metrics
#'
#' @param dm matrix
#' 
#' @useDynLib dataAnalysisMisc
#'
#' @export
getMF <- function(dm) {
    mode(dm) <- "integer"    
    n <- integer(length(dm[1,]))       
    cons <- integer(length(dm[1,]))       
    len1 <- as.integer(length(dm[,1]))    
    len2 <- as.integer(length(dm[1,]))    
    katMax <- as.integer(max(dm)+1)    

    ret <- .C("cons", matrix=dm, len1=len1, len2=len2, sel=cons, n=n, katMax=katMax)    
    return(list(ret$sel, ret$n))
}


#' @title Match and subst seq
#' 
#' @description ... 
#'
#' @param seq seq
#' @param search pattern
#' 
#' @useDynLib dataAnalysisMisc
#'
#' @export
matchSub <- function(seq, search) {
    seqN <- integer(nchar(seq))    
    len <- as.integer(nchar(seq))    
    posM <- as.integer(1)

    ret <- .C("mtchSub", len=len, pattern=search, string=seq, seqN=seqN, posM=posM)    
    #ret    
    found <- any(ret$seqN != 0)

    if (found) {
	return(list(found=found, num=ret$seqN[-c(1:ret$posM)]))
    } else {
	return(list(found=found, num=ret$seqN))
    }
}


#' @title Match and subst seq rev
#' 
#' @description ... 
#'
#' @param seq seq
#' @param search pattern
#' 
#' @useDynLib dataAnalysisMisc
#'
#' @export
matchSubRev <- function(seq, search) {
    seqN <- integer(nchar(seq))    
    len <- as.integer(nchar(search))    
    posM <- as.integer(1)

    ret <- .C("mtchSubRev", len=len, pattern=search, string=seq, seqN=seqN, posM=posM)    
    #ret    
    found <- any(ret$seqN != 0)

    if (found) {
	return(list(found=found, num=rev(ret$seqN[-c((ret$posM+nchar(search)):length(ret$seqN))])))
    } else {
	warning("Not implmeneted yet!")
	#return(list(found=found, num=ret$seqN))
    }
}


#' @title Test occurence of substring
#' 
#' @description ... 
#'
#' @param seq seq
#' @param search pattern
#' 
#' @useDynLib dataAnalysisMisc
#'
#' @export
matchPattern <- function(search, seq) {
    res <- integer(1)

    ret <- .C("mtch", pattern=search, string=seq, res=res)    
    return (ret$res)
}
